tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_UpwindFlux.cpp

```diff
@@ -0,0 +1,297 @@
+  packaged_data_ext_opposite_normal(
+      get(one).size(), std::numeric_limits<double>::signaling_NaN());
+  flux_computer.package_data(make_not_null(&packaged_data_ext_opposite_normal),
+                             spacetime_metric_ext, pi_ext, phi_ext, lapse_ext,
+                             shift_ext, inverse_spatial_metric_ext, gamma_1,
+                             gamma_2, minus_unit_normal_one_form_ext);
+
+  // Check that if the same fields are given for the interior and exterior
+  // (except that the normal vector gets multiplied by -1.0) that the
+  // numerical flux reduces to the flux
+  auto psi_normal_dot_numerical_flux = make_with_value<
+      db::item_type<::Tags::NormalDotNumericalFlux<gr::Tags::SpacetimeMetric<
+          spatial_dim, Frame::Inertial, DataVector>>>>(one, 0.0);
+  auto pi_normal_dot_numerical_flux =
+      make_with_value<db::item_type<::Tags::NormalDotNumericalFlux<
+          GeneralizedHarmonic::Tags::Pi<spatial_dim, Frame::Inertial>>>>(one,
+                                                                         0.0);
+  auto phi_normal_dot_numerical_flux =
+      make_with_value<db::item_type<::Tags::NormalDotNumericalFlux<
+          GeneralizedHarmonic::Tags::Phi<spatial_dim, Frame::Inertial>>>>(one,
+                                                                          0.0);
+  ::TestHelpers::NumericalFluxes::apply_numerical_flux(
+      flux_computer, packaged_data_int, packaged_data_int_opposite_normal,
+      make_not_null(&psi_normal_dot_numerical_flux),
+      make_not_null(&pi_normal_dot_numerical_flux),
+      make_not_null(&phi_normal_dot_numerical_flux));
+
+  ::GeneralizedHarmonic::ComputeNormalDotFluxes<spatial_dim>
+      normal_dot_flux_computer{};
+  auto psi_normal_dot_flux = make_with_value<
+      db::item_type<::Tags::NormalDotNumericalFlux<gr::Tags::SpacetimeMetric<
+          spatial_dim, Frame::Inertial, DataVector>>>>(one, 0.0);
+  auto pi_normal_dot_flux =
+      make_with_value<db::item_type<::Tags::NormalDotNumericalFlux<
+          GeneralizedHarmonic::Tags::Pi<spatial_dim, Frame::Inertial>>>>(one,
+                                                                         0.0);
+  auto phi_normal_dot_flux =
+      make_with_value<db::item_type<::Tags::NormalDotNumericalFlux<
+          GeneralizedHarmonic::Tags::Phi<spatial_dim, Frame::Inertial>>>>(one,
+                                                                          0.0);
+  normal_dot_flux_computer.apply(
+      make_not_null(&psi_normal_dot_flux), make_not_null(&pi_normal_dot_flux),
+      make_not_null(&phi_normal_dot_flux), spacetime_metric_int, pi_int,
+      phi_int, gamma_1, gamma_2, lapse_int, shift_int,
+      inverse_spatial_metric_int, unit_normal_one_form_int);
+
+  CHECK_ITERABLE_APPROX(psi_normal_dot_numerical_flux, psi_normal_dot_flux);
+  CHECK_ITERABLE_APPROX(pi_normal_dot_numerical_flux, pi_normal_dot_flux);
+  CHECK_ITERABLE_APPROX(phi_normal_dot_numerical_flux, phi_normal_dot_flux);
```

Comment:
Please add some structure to the code to explain what is being tested using scopes and `INFO`:
```cpp
{
     INFO(...);
}
```

Here you only test the consistency of the flux, i.e. F^i_num (u,u) = F^i(u) .
What is missing is conservation, i.e. the numerical flux is single valued when computed from either side. [[ F^i_num (u_int, n_int, u_ext, n_ext) ]]_i = n_i^int F^i_num(u_int, n_int, u_ext, n_ext) + n_i^ext F^i_num (u_ext, n_ext, u_int, n_int) = 0 .

Reply:
The GH upwind flux as implemented does not satisfy conservation, as explained in its help text's warning. So I only test consistency here.

The plan (discussed earlier in this thread and on a call involving @nilsdeppe, @wthrowe, and others) is to first merge this flux as is. Then, I will look into alternatives, such as the one @nilsdeppe suggested, that do satisfy conservation. Since this flux is only intended for use on smooth solutions (i.e. vacuum), I suspect that it will work fine in general, though so far I have only tested it on KerrSchild.

Reply:
If you only test consistency, then you don't need the code that sets up the second randomly generated set of variables (you call them `_ext`). Please add a comment on what you are testing to the code nonetheless, and that you are not testing conservation.

Reply:
I test consistency, but part of that test is that the weighting works properly, i.e. only returns the int or ext fields if the char speeds are + or -. So I do need the _ext quantities. But I do now comment that I'm testing consistency only.